functions/v2/ocr/app/index.js

```diff
@@ -0,0 +1,222 @@
+};
+// [END functions_ocr_detect]
+
+/**
+ * Appends a .txt suffix to the image name.
+ *
+ * @param {string} filename Name of a file.
+ * @param {string} lang Language to append.
+ * @returns {string} The new filename.
+ */
+const renameImageForSave = (filename, lang) => {
+  return `${filename}_to_${lang}.txt`;
+};
+
+// [START functions_ocr_process]
+/**
+ * This function is exported by index.js, and is executed when
+ * a file is uploaded to the Cloud Storage bucket you created
+ * for uploading images.
+ *
+ * @param {object} cloudEvent A CloudEvent containing the Cloud Storage File object.
+ */
+functions.cloudEvent('processImage', async cloudEvent => {
+  const {bucket, name} = cloudEvent.data;
+
+  if (!bucket) {
+    throw new Error(
+      'Bucket not provided. Make sure you have a "bucket" property in your request'
+    );
+  }
+  if (!name) {
+    throw new Error(
+      'Filename not provided. Make sure you have a "name" property in your request'
+    );
+  }
+
+  const results = await detectText(bucket, name);
+  console.log(`File ${name} processed.`);
+  return results;
+});
+// [END functions_ocr_process]
+
+// [START functions_ocr_translate]
+/**
+ * This function is exported by index.js, and is executed when
+ * a message is published to the Cloud Pub/Sub topic specified
+ * by the TRANSLATE_TOPIC environment variable. The function
+ * translates text using the Google Translate API.
+ *
+ * @param {object} cloudEvent The CloudEvent containing the Pub/Sub Message object
+ */
+functions.cloudEvent('translateText', async cloudEvent => {
+  const pubsubData = cloudEvent.data;
+  const jsonStr = Buffer.from(pubsubData.message, 'base64').toString();
+  const {text, filename, lang} = JSON.parse(jsonStr);
+
+  if (!text) {
+    throw new Error(
+      'Text not provided. Make sure you have a "text" property in your request'
+    );
+  }
+  if (!filename) {
+    throw new Error(
+      'Filename not provided. Make sure you have a "filename" property in your request'
+    );
+  }
+  if (!lang) {
+    throw new Error(
+      'Language not provided. Make sure you have a "lang" property in your request'
+    );
+  }
+
+  console.log(`Translating text into ${lang}`);
+  const [translation] = await translate.translate(text, lang);
+
+  console.log('Translated text:', translation);
+
+  const messageData = {
+    text: translation,
+    filename: filename,
+    lang: lang,
+  };
+
+  await publishResult(process.env.RESULT_TOPIC, messageData);
+  console.log(`Text translated to ${lang}`);
+});
+// [END functions_ocr_translate]
+
+// [START functions_ocr_save]
+/**
+ * This function is exported by index.js, and is executed when
+ * a message is published to the Cloud Pub/Sub topic specified
+ * by the RESULT_TOPIC environment variable. The function saves
+ * the data packet to a file in GCS.
+ *
+ * @param {object} cloudEvent The CloudEvent containing the Pub/Sub Message object.
+ */
+functions.cloudEvent('saveResult', async cloudEvent => {
```

Comment:
Should each function be in its own directory?

(For one thing, I think each function has a different set of dependencies.)

Reply:
Good question, I'm not sure, but I'll look into it.

Reply:
Also worth noting most of this code is not new, it's just slightly modified from here:
https://github.com/GoogleCloudPlatform/nodejs-docs-samples/tree/main/functions/ocr

But if you think there's changes worth making here, I'm open to doing so.

Reply:
Clarified this internally

**TL;DR:** this is fine as-is. 🙂